feat: Add ignoreOrder flag

mjhenkes opened the pull request:
<!--
  HOLY CRAP a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->

This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case
Picking up the torch where https://github.com/webpack-contrib/mini-css-extract-plugin/pull/366 left off and making the changes suggested by @evilebottnawi 
Also addresses #362
Borrowing @hedgepigdaniel 's description:

As has been pointed out in #250:

* Sorting all imports (not just in files but in tree traversal import order is difficult to do in a large project, and impossible if there are async chunks with CSS.
* Consistent use of scoping or naming conventions in CSS makes the order of CSS import irrelevant, and this is a much more manageable strategy for a large project as @jsg2021 said in #250 (comment)
* Using stats.warningFilter as recommended by @sokra in #250 (comment) or webpack-filter-warnings-plugin does not work consistently. It's much better to avoid generating the warnings at their source (optionally) rather than to partially filter them later with regexes. For example, I use webpack-error-on-warnings-plugin to enforce that there are no warnings in a build. This is incompatible with filtering warnings, because it the error must happen in the before-emit hook so that it prevents emission of assets when there are warnings. The warnings filtering happens in the done hook or in the toJson method, which is too late.

<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->

### Breaking Changes
None
### Additional Info
The added tests is to ensure that enabling the flag doesn't make the plugin implode, but it doesn't really programmatically verify the warnings aren't printed.


fixes #362

kubijo commented:
Can someone from mainteiners merge this please.
The code seems nice, terse and without conflicts, so this should be a quick pass-over

mjhenkes commented:
@evilebottnawi technically all the other tests are testing ignoreOrder: false, but i've added an explicit test.
I've also updated the package-lock file, it started failing due to npm audit.

mjhenkes commented:
@evilebottnawi added

evilebottnawi commented on test/IgnoreOrder.test.js:
```diff
@@ -0,0 +1,44 @@
+      cache: false,
+    });
+    compiler.run((err1, stats) => {
+      expect(stats.hasWarnings()).toBeTruthy();
```
`toBe(true)`

JimTheMan replied:
just wondering why toBeTruthy(); is "better". Seems like the exact same shit to me...

JimTheMan replied:
@evilebottnawi

evilebottnawi commented on test/IgnoreOrder.test.js:
```diff
@@ -0,0 +1,44 @@
+      cache: false,
+    });
+    compiler.run((err1, stats) => {
+      expect(stats.hasWarnings()).toBeFalsy();
```
`toBe(false)`

evilebottnawi commented on test/IgnoreOrder.test.js:
```diff
@@ -0,0 +1,44 @@
+import path from 'path';
```
rename file `ignoreOrder-option.test.js`

mjhenkes commented:
Do we need more approvals?

mjhenkes commented:
Thanks @evilebottnawi! 🎉

kubijo commented:
Yeah, thanks a lot guys!
What is the release process now?

evilebottnawi commented:
Today
